### PR TITLE
feat: added favourites

### PR DESCRIPTION
**Pull request by Yian**
-- Ability to add or remove activity from favourites
-- Clicking favourites in UI will take you to /favourites link and show favourite activities
-- Styling for Avatar menu

dev notes:
-- fixed tests 
-- cleanup 
-- refactoring and renaming

### CODE REVIEW

File: src/app/app.component.ts [hunk 1]
```diff
@@ -40,7 +39,7 @@ export class AppComponent implements OnInit {
       callback: () => {
         this.analyticsService.logEvent('email_developer');
         window.open(
-          'mailto:azhidkov@gmail.com?subject=ScreenHush%20App&body=Hi%20Alex,%20Love%20your%20app!'
+          'mailto:azhidkov@gmail.com?subject=Team%20Builder%20App&body=Hi%20Alex,%20Love%20your%20app!'
```
Comment (AlexZhidkov) [inline src/app/app.component.ts:42]: Reverted my change. I'll fix it myself.

File: src/app/services/favourites.service.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+          console.error('Error retrieving favorite activities:', error);
+        }
+      } else {
+        console.error('User UID is null');
+      }
+    } else {
+      console.error('User not logged in');
+    }
+
+    return []; // Return an empty array if there is an error or user is not logged in
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts]: I like to do all checks up top at the start of the function and return this if they fail. I think it's more readable with less indentations.

File: src/app/services/favourites.service.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+          // Retrieve detailed information about the activities from the "activities" collection
+          const favouriteActivities = await Promise.all(
+            activityIds.map(async (activityId) => {
+              const activityDocRef = doc(this.firestore, 'activities', activityId);
+              const activityDocSnapshot = await getDoc(activityDocRef);
+
+              if (activityDocSnapshot.exists()) {
+                return { id: activityId, ...activityDocSnapshot.data() };
+              } else {
+                console.warn(`Activity with ID ${activityId} not found`);
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts]: In this case, I would delete it from the favourites collection.

File: src/app/services/favourites.service.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    } else {
+      console.error('User not logged in');
+    }
+
+    return []; // Return an empty array if there is an error or user is not logged in
+  }
+
+  async addToFavourites(activityId: string | null) {
+    if (this.currentUser) {
+      const queryRef = query(this.favouritesCollection, where('activityId', '==', activityId));
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts]: Do you think the cost of querying is justified? I wouldn't do it.

File: src/app/services/favourites.service.ts (hunk 2 above)
Comment (Yian) [reply src/app/services/favourites.service.ts]: I'll update this

File: src/app/services/favourites.service.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        await addDoc(this.favouritesCollection, { activityId: activityId });
+      }
+    } else {
+      console.error('User not logged in');
+    }
+  }
+
+  async removeFromFavourites(activityId: string | null) {
+    if (this.currentUser) {
+      const queryRef = query(this.favouritesCollection, where('activityId', '==', activityId));
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts]: Same here, Can we just delete without confirming it exists? Would we save some on these calls?

File: src/app/services/favourites.service.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+      const querySnapshot = await getDocs(queryRef);
+  
+      if (!querySnapshot.empty) {
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts]: Is `querySnapshot.exists()` better?

File: src/app/services/favourites.service.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+      // Update favouritesCollection when the authentication state changes
+      const userUid = this.currentUser?.uid;
+
+      this.favouritesCollection = userUid
+      ? collection(this.firestore, 'users', userUid, 'favourites')
+      : (null as unknown as CollectionReference);
+    });
+  }
+
+  async isActivityFavorited(activityId: string | null): Promise<boolean> {
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts:44]: Can we have a local in-memory array of favourites instead of querying db for every activity?

Comment (AlexZhidkov) [reply src/app/services/favourites.service.ts:44]: We can load it whenever the user is logged in and maintain in-memory on add/remove.

Comment (Yian) [reply src/app/services/favourites.service.ts:44]: I've removed the activity query, we can just check in the users array now

File: src/app/services/favourites.service.ts (hunk 3 above)
Comment (Yian) [reply src/app/services/favourites.service.ts]: Have added this

File: src/app/services/favourites.service.ts (hunk 4 above)
Comment (Yian) [reply src/app/services/favourites.service.ts]: have updated to use the array method

File: src/app/services/favourites.service.ts (hunk 5 above)
Comment (Yian) [reply src/app/services/favourites.service.ts]: will just remove from array now

File: src/app/services/favourites.service.ts (hunk 6 above)
Comment (Yian) [reply src/app/services/favourites.service.ts]: removed this one

File: src/app/services/favourites.service.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    
+  }
+
+  async isActivityFavorited(activityId: string | null): Promise<boolean> {
+    if (!this.currentUser && !activityId) {
+      throw new Error('User not logged in or activityId is null');
+    }
+
+    const userDocSnapshot = await getDoc(this.userDocRef);
+    const favorites = userDocSnapshot.get('favorites') || [];
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts]: Can we have this array in memory instead of making a request for every activity?

Comment (AlexZhidkov) [reply src/app/services/favourites.service.ts]: If you have `favorites` array as a property, load it in constractor and update in add/remove functions?

Comment (Yian) [reply src/app/services/favourites.service.ts]: Done

File: src/app/services/favourites.service.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    this.userFavoritesCache = (await userDocSnapshot).get('favorites') || [];
+  }
+
+  async isActivityFavorited(activityId: string | null): Promise<boolean> {
+    if (!this.currentUser || activityId === null) {
+      throw new Error('User not logged in or activityId is null');
+    }
+
+    if (!this.userFavoritesCache || this.userFavoritesCache.length === 0) {
+      await this.loadFavouritesInCache();
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts]: So if user doesn't have any activities favorited, it's going to keep loading for every activity. Can we remove this?

File: src/app/services/favourites.service.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+})
+export class FavouritesService {
+  private firestore: Firestore = inject(Firestore);
+  private currentUser: User | null;
+  private userDocRef: DocumentReference<DocumentData>;
+  private userFavoritesCache: string[];
+
+  constructor(private authService: AuthenticationService) {
+    this.setupAuthStatusListener();
+    this.loadFavouritesInCache();
```
Comment (AlexZhidkov) [inline src/app/services/favourites.service.ts]: We need to have user's auth status before loading his favorites, right? Do we need to await for auth?

File: src/app/services/favourites.service.ts (hunk 9 above)
Comment (Yian) [reply src/app/services/favourites.service.ts]: Yeah this is not actually needed, have removed

File: src/app/services/favourites.service.ts (hunk 10 above)
Comment (Yian) [reply src/app/services/favourites.service.ts]: Loaded after auth obtained

**Review by AlexZhidkov — APPROVED** [review]

**Pull request merged**